File: algorithm_direct/direct_rme_algorithm.py
"""
Simple standalone Random Mask Encoding phase mask code.

This is plain Python reference code with no UI dependency.

Dependencies:
    pip install numpy pillow
"""

from pathlib import Path

import numpy as np
from PIL import Image


TWO_PI = 2.0 * np.pi
SEED_BASE = 0x2C9277B5E57A4A83


def default_settings():
    """All needed variables in one simple dictionary."""
    return {
        # SLM hardware
        "slm_width": 1920,
        "slm_height": 1080,
        "slm_pixel_size_um": 8.0,

        # Camera / trap plane
        "cam_width": 1920,
        "cam_height": 1080,
        "cam_pixel_size_um": 5.0,
        "camera_imaging_magnification": 1.0,

        # Optical setup
        "wavelength_nm": 1064.0,
        "focal_length_mm": 100.0,
    }


def mapping_values(settings):
    """Compute physical camera pixel pitch and focal-plane FFT-grid pitch."""
    slm_dx = settings["slm_pixel_size_um"] * 1e-6
    slm_dy = settings["slm_pixel_size_um"] * 1e-6

    effective_cam_pixel_um = (
        settings["cam_pixel_size_um"] / settings["camera_imaging_magnification"]
    )
    cam_dx = effective_cam_pixel_um * 1e-6
    cam_dy = effective_cam_pixel_um * 1e-6

    wavelength = settings["wavelength_nm"] * 1e-9
    focal_length = settings["focal_length_mm"] * 1e-3

    focal_dx = wavelength * focal_length / (settings["slm_width"] * slm_dx)
    focal_dy = wavelength * focal_length / (settings["slm_height"] * slm_dy)
    return cam_dx, cam_dy, focal_dx, focal_dy


def map_points_to_fft_offsets(points, settings):
    """
    Convert camera-centered trap points to Fourier-grid offsets.

    Point coordinates are camera-centered:
    (0, 0) is center, +x is right, +y is up.
    """
    width = settings["slm_width"]
    height = settings["slm_height"]
    cam_width = settings["cam_width"]
    cam_height = settings["cam_height"]

    cam_dx, cam_dy, focal_dx, focal_dy = mapping_values(settings)
    cx = width // 2
    cy = height // 2

    offsets = set()
    skipped_camera = 0
    skipped_slm = 0

    for x_cam, y_cam in points:
        if abs(x_cam) > cam_width / 2.0 or abs(y_cam) > cam_height / 2.0:
            skipped_camera += 1
            continue

        fft_offset_x = int(np.rint((x_cam * cam_dx) / focal_dx))
        fft_offset_y = int(np.rint((y_cam * cam_dy) / focal_dy))

        fft_x = cx + fft_offset_x
        fft_y = cy - fft_offset_y

        if fft_x < 0 or fft_x >= width or fft_y < 0 or fft_y >= height:
            skipped_slm += 1
            continue

        offsets.add((fft_offset_x, fft_offset_y))

    info = {
        "requested": len(points),
        "used": len(offsets),
        "skipped_camera": skipped_camera,
        "skipped_slm": skipped_slm,
    }
    return sorted(offsets), info


def seed_mix(seed, value):
    """Small deterministic seed mixer, matching the C++ idea."""
    seed ^= value + 0x9E3779B97F4A7C15 + ((seed << 6) & 0xFFFFFFFFFFFFFFFF) + (seed >> 2)
    return seed & 0xFFFFFFFFFFFFFFFF


def deterministic_seed(width, height, offsets):
    """Build a deterministic random seed from the target geometry."""
    seed = SEED_BASE
    seed = seed_mix(seed, width)
    seed = seed_mix(seed, height)
    seed = seed_mix(seed, len(offsets))

    for ox, oy in offsets:
        seed = seed_mix(seed, int(ox) & 0xFFFFFFFFFFFFFFFF)
        seed = seed_mix(seed, int(oy) & 0xFFFFFFFFFFFFFFFF)

    out = ((seed >> 32) ^ (seed & 0xFFFFFFFF)) & 0xFFFFFFFF
    return out if out != 0 else 0xA511E9B3


def phase_to_8bit(phase):
    """Wrap phase to [0, 2pi) and convert it to 8-bit grayscale."""
    wrapped = np.mod(phase, TWO_PI)
    mask = np.clip((wrapped / TWO_PI) * 255.0, 0, 255).astype(np.uint8)
    return mask, wrapped


def random_mask_encoding(points, settings=None):
    """
    Generate a Random Mask Encoding phase mask.

    Idea:
    1. Map every target point to a Fourier-grid offset.
    2. Randomly assign each SLM pixel to one target.
    3. For each SLM pixel, write the grating phase for its assigned target.
    """
    if settings is None:
        settings = default_settings()

    width = settings["slm_width"]
    height = settings["slm_height"]
    pixel_count = width * height

    offsets, info = map_points_to_fft_offsets(points, settings)
    if not offsets:
        raise ValueError("no valid target points after camera/Fourier mapping")

    trap_count = len(offsets)
    rng = np.random.default_rng(deterministic_seed(width, height, offsets))

    # Shuffle all SLM pixels and split them as evenly as possible across traps.
    pixel_order = np.arange(pixel_count)
    rng.shuffle(pixel_order)

    trap_for_pixel = np.zeros(pixel_count, dtype=int)
    cursor = 0
    base_count = pixel_count // trap_count
    remainder = pixel_count % trap_count

    for trap_index in range(trap_count):
        quota = base_count + (1 if trap_index < remainder else 0)
        chosen_pixels = pixel_order[cursor: cursor + quota]
        trap_for_pixel[chosen_pixels] = trap_index
        cursor += quota

    trap_for_pixel = trap_for_pixel.reshape(height, width)

    y, x = np.indices((height, width), dtype=float)
    offset_x = np.array([p[0] for p in offsets], dtype=float)[trap_for_pixel]
    offset_y = np.array([p[1] for p in offsets], dtype=float)[trap_for_pixel]

    phase = TWO_PI * ((offset_x * x / width) - (offset_y * y / height))
    mask, wrapped_phase = phase_to_8bit(phase)

    return {
        "phase_mask_8bit": mask,
        "wrapped_phase_rad": wrapped_phase,
        "info": info,
    }


def save_mask(path, phase_mask_8bit):
    """Save an 8-bit grayscale phase mask."""
    Image.fromarray(np.asarray(phase_mask_8bit, dtype=np.uint8), mode="L").save(Path(path))


if __name__ == "__main__":
    settings = default_settings()
    result = random_mask_encoding([(0.0, 0.0), (100.0, 0.0)], settings)
    save_mask("rme_mask.png", result["phase_mask_8bit"])

File: algorithm_direct/direct_rme_algorithm.cpp
#include "direct_rme_algorithm.h"

#include <algorithm>
#include <cmath>
#include <cstdint>
#include <numeric>
#include <random>
#include <set>
#include <utility>
#include <vector>

namespace {
constexpr double kTwoPi = 6.28318530717958647692;
constexpr std::uint64_t kSeedBase = 0x2C9277B5E57A4A83ull;

double wrapPhaseRad(double phase) {
    double wrapped = std::fmod(phase, kTwoPi);
    if (wrapped < 0.0) {
        wrapped += kTwoPi;
    }
    return wrapped;
}

void seedMix(std::uint64_t &seed, std::uint64_t value) {
    seed ^= value + 0x9e3779b97f4a7c15ull + (seed << 6u) + (seed >> 2u);
}

std::uint32_t buildDeterministicSeed(int width,
                                     int height,
                                     const std::vector<std::pair<int, int>> &targetOffsets) {
    std::uint64_t seed = kSeedBase;
    seedMix(seed, static_cast<std::uint64_t>(width));
    seedMix(seed, static_cast<std::uint64_t>(height));
    seedMix(seed, static_cast<std::uint64_t>(targetOffsets.size()));
    for (const auto &offset : targetOffsets) {
        seedMix(seed, static_cast<std::uint64_t>(static_cast<std::int64_t>(offset.first)));
        seedMix(seed, static_cast<std::uint64_t>(static_cast<std::int64_t>(offset.second)));
    }

    std::uint32_t out = static_cast<std::uint32_t>((seed >> 32u) ^ (seed & 0xffffffffu));
    if (out == 0u) {
        out = 0xA511E9B3u;
    }
    return out;
}

bool prepareMappedOffsets(const AlgorithmDirect::DirectRmeVariables &variables,
                          const QVector<AlgorithmDirect::DirectRmeTargetPoint> &targets,
                          AlgorithmDirect::DirectRmeResult &result,
                          std::vector<std::pair<int, int>> &mappedOffsets) {
    result = AlgorithmDirect::DirectRmeResult();
    result.requestedTargetCount = targets.size();

    if (variables.slmWidth <= 0 || variables.slmHeight <= 0) {
        result.error = "Invalid SLM resolution for RME algorithm.";
        return false;
    }
    if (variables.camWidth <= 0 || variables.camHeight <= 0) {
        result.error = "Invalid camera resolution for RME algorithm.";
        return false;
    }
    if (variables.slmPixelSizeUm <= 0.0 || variables.camPixelSizeUm <= 0.0 ||
        variables.cameraImagingMagnification <= 0.0 ||
        variables.wavelengthNm <= 0.0 || variables.focalLengthMm <= 0.0) {
        result.error = "Optical/hardware settings must be positive before running RME.";
        return false;
    }
    if (targets.isEmpty()) {
        result.error = "No target points found. Add at least one point on the target grid.";
        return false;
    }

    const double slmDx = variables.slmPixelSizeUm * 1e-6;
    const double slmDy = variables.slmPixelSizeUm * 1e-6;
    const double effectiveCamPixelSizeUm = variables.camPixelSizeUm / variables.cameraImagingMagnification;
    const double camDx = effectiveCamPixelSizeUm * 1e-6;
    const double camDy = effectiveCamPixelSizeUm * 1e-6;
    const double wavelength = variables.wavelengthNm * 1e-9;
    const double focalLength = variables.focalLengthMm * 1e-3;

    const double focalDx = (wavelength * focalLength) / (static_cast<double>(variables.slmWidth) * slmDx);
    const double focalDy = (wavelength * focalLength) / (static_cast<double>(variables.slmHeight) * slmDy);
    if (focalDx == 0.0 || focalDy == 0.0) {
        result.error = "Computed focal-plane pixel pitch is zero. Check optical settings.";
        return false;
    }

    const int cx = variables.slmWidth / 2;
    const int cy = variables.slmHeight / 2;

    std::set<std::pair<int, int>> dedupOffsets;
    for (const AlgorithmDirect::DirectRmeTargetPoint &target : targets) {
        if (std::abs(target.xCamPx) > static_cast<double>(variables.camWidth) / 2.0 ||
            std::abs(target.yCamPx) > static_cast<double>(variables.camHeight) / 2.0) {
            ++result.skippedOutsideCameraFov;
            continue;
        }

        const double physX = target.xCamPx * camDx;
        const double physY = target.yCamPx * camDy;
        const int fftOffsetX = static_cast<int>(std::llround(physX / focalDx));
        const int fftOffsetY = static_cast<int>(std::llround(physY / focalDy));

        const int fftX = cx + fftOffsetX;
        const int fftY = cy - fftOffsetY;
        if (fftX < 0 || fftX >= variables.slmWidth || fftY < 0 || fftY >= variables.slmHeight) {
            ++result.skippedOutsideSlmBounds;
            continue;
        }

        dedupOffsets.insert({fftOffsetX, fftOffsetY});
    }

    if (dedupOffsets.empty()) {
        result.error = "No valid target points remained after camera/Fourier mapping.";
        return false;
    }

    mappedOffsets.assign(dedupOffsets.begin(), dedupOffsets.end());
    result.usedTargetCount = static_cast<int>(mappedOffsets.size());
    return true;
}
} // namespace

namespace AlgorithmDirect {

QVector<DirectRmeTargetPoint> buildRmePointTargets(const QVector<QPointF> &cameraCenteredPoints) {
    QVector<DirectRmeTargetPoint> targets;
    targets.reserve(cameraCenteredPoints.size());

    for (const QPointF &point : cameraCenteredPoints) {
        DirectRmeTargetPoint target;
        target.xCamPx = point.x();
        target.yCamPx = point.y();
        targets.append(target);
    }

    return targets;
}

DirectRmeResult runRandomMaskEncoding(const DirectRmeVariables &variables,
                                      const QVector<DirectRmeTargetPoint> &targets) {
    DirectRmeResult result;
    std::vector<std::pair<int, int>> targetOffsets;
    if (!prepareMappedOffsets(variables, targets, result, targetOffsets)) {
        return result;
    }

    const int width = variables.slmWidth;
    const int height = variables.slmHeight;
    const int pixelCount = width * height;
    const int trapCount = static_cast<int>(targetOffsets.size());
    if (pixelCount <= 0 || trapCount <= 0) {
        result.error = "Invalid RME internal dimensions.";
        return result;
    }

    // Deterministically assign SLM pixels to target traps in random order.
    std::vector<int> pixelOrder(static_cast<std::size_t>(pixelCount));
    std::iota(pixelOrder.begin(), pixelOrder.end(), 0);
    std::mt19937 rng(buildDeterministicSeed(width, height, targetOffsets));
    std::shuffle(pixelOrder.begin(), pixelOrder.end(), rng);

    std::vector<int> trapForPixel(static_cast<std::size_t>(pixelCount), 0);
    int cursor = 0;
    const int baseCount = pixelCount / trapCount;
    const int remainder = pixelCount % trapCount;

    for (int trapIndex = 0; trapIndex < trapCount; ++trapIndex) {
        const int quota = baseCount + ((trapIndex < remainder) ? 1 : 0);
        for (int j = 0; j < quota && cursor < pixelCount; ++j) {
            trapForPixel[static_cast<std::size_t>(pixelOrder[static_cast<std::size_t>(cursor)])] = trapIndex;
            ++cursor;
        }
    }

    result.wrappedPhaseRad.resize(pixelCount);
    result.phaseMask8Bit = QImage(width, height, QImage::Format_Grayscale8);
    if (result.phaseMask8Bit.isNull()) {
        result.error = "Failed to allocate RME phase mask image.";
        return result;
    }

    const double invWidth = 1.0 / static_cast<double>(width);
    const double invHeight = 1.0 / static_cast<double>(height);
    for (int y = 0; y < height; ++y) {
        uchar *row = result.phaseMask8Bit.scanLine(y);
        const int rowBase = y * width;
        for (int x = 0; x < width; ++x) {
            const int idx = rowBase + x;
            const int trapIndex = trapForPixel[static_cast<std::size_t>(idx)];
            const auto &offset = targetOffsets[static_cast<std::size_t>(trapIndex)];

            const double phase = kTwoPi * (static_cast<double>(offset.first) * static_cast<double>(x) * invWidth -
                                           static_cast<double>(offset.second) * static_cast<double>(y) * invHeight);
            const double wrapped = wrapPhaseRad(phase);
            result.wrappedPhaseRad[idx] = wrapped;

            const double scaled = (wrapped / kTwoPi) * 255.0;
            row[x] = static_cast<uchar>(std::clamp(static_cast<int>(scaled), 0, 255));
        }
    }

    result.success = true;
    return result;
}

DirectRmeResult generateRandomMaskEncodingPhaseMask(const QVector<QPointF> &cameraCenteredPoints,
                                                    const DirectRmeVariables &variables) {
    DirectRmeResult result;
    if (cameraCenteredPoints.isEmpty()) {
        result.error = "No camera-centered target points were provided.";
        return result;
    }

    return runRandomMaskEncoding(variables, buildRmePointTargets(cameraCenteredPoints));
}

} // namespace AlgorithmDirect
